daemon/session.c

```diff
@@ -0,0 +1,354 @@
+
+uid_t
+session_current_user(const session_t *self)
+{
+    return self->ssn_active_uid;
+}
+
+static void
+session_notify_changed(const session_t *self)
+{
+    if( self->ssn_initialized ) {
+        log_info("SESSION MONITOR: notify");
+        control_on_session_changed(session_control(self));
+    }
+}
+
+/* ========================================================================= *
+ * SESSION_MONITOR
+ * ========================================================================= */
+
+/** Handle systemd login monitoring wakeups
+ */
+static gboolean
+session_monitor_cb(GIOChannel *chn, GIOCondition cnd, gpointer aptr)
+{
+    (void)chn;
+    session_t *self = aptr;
+    gboolean result = G_SOURCE_REMOVE;
+
+    if( !self->ssn_monitor_id || !self->ssn_monitor_obj )
+        goto EXIT;
+
+    if( cnd & ~G_IO_IN )
+        goto EXIT;
+
+    session_update_monitor(self);
+
+    if( sd_login_monitor_flush(self->ssn_monitor_obj) >= 0 )
+        result = G_SOURCE_CONTINUE;
+
+EXIT:
+    if( result == G_SOURCE_REMOVE && self->ssn_monitor_id ) {
+        log_crit("SESSION MONITOR: disabled");
+        self->ssn_monitor_id = 0;
+        session_stop_monitor(self);
+    }
+
+    return result;
+}
+
+/** Start user session monitoring
+ */
+static void
+session_start_monitor(session_t *self)
+{
+    bool ack = false;
+
+    session_stop_monitor(self);
+
+    if( sd_login_monitor_new("session", &self->ssn_monitor_obj) < 0 )
+        goto EXIT;
+
+    int fd = sd_login_monitor_get_fd(self->ssn_monitor_obj);
+    if( fd < 0 )
+        goto EXIT;
+
+    self->ssn_monitor_id = gutil_add_watch(fd, G_IO_IN,
+                                           session_monitor_cb, self);
+    if( !self->ssn_monitor_id )
+        goto EXIT;
+
+    ack = true;
+
+    log_info("SESSION MONITOR: started");
+
+EXIT:
+    if( !ack )
+        session_stop_monitor(self);
+
+    return;
+}
+
+/** Stop user session monitoring
+ */
+static void
+session_stop_monitor(session_t *self)
+{
+    self->ssn_control = NULL;
+
+    if( self->ssn_monitor_id ) {
+        log_info("SESSION MONITOR: stopped");
+
+        g_source_remove(self->ssn_monitor_id),
+            self->ssn_monitor_id = 0;
+    }
+
+    if( self->ssn_monitor_obj ) {
+        sd_login_monitor_unref(self->ssn_monitor_obj),
+            self->ssn_monitor_obj = 0;
+    }
+}
+
+static void
+session_update_monitor(session_t *self)
+{
+    uid_t uid = session_get_seat0_uid();
+    if( self->ssn_active_uid != uid ) {
+        log_info("SESSION MONITOR: uid: %d -> %d",
+                  (int)self->ssn_active_uid,
+                  (int)uid);
+        self->ssn_active_uid = uid;
+        session_notify_changed(self);
+    }
+}
+
+/* ========================================================================= *
+ * SESSION_SEAT
+ * ========================================================================= */
+
+/** Find uid for user active session at seat0
+ */
+static uid_t
+session_get_seat0_uid(void)
```

Comment:
Why can't we ask usermanagerd about the current user?

Reply:
Because I did not spot that user manager does provide user enumeration. Which is why it is not used. And using it would change everything. Which is why it is handled as a separate task. And this applies for everything related to user manager.